azurelinuxagent/common/persist_firewall_rules.py

```diff
@@ -0,0 +1,235 @@
+            self._setup_permanent_firewalld_rules()
+            return
+
+        logger.info(
+            "Firewalld service not running/unavailable, trying to set up {0}".format(self._network_setup_service_name))
+
+        if not CGroupsApi.is_systemd():
+            raise Exception("Systemd not enabled, unable to set {0}".format(self._network_setup_service_name))
+
+        self._setup_network_setup_service()
+
+    def __verify_firewall_rules_enabled(self):
+        # Check if firewall-rules have already been enabled
+        # This function would also return False if the dest-ip is changed. So no need to check separately for that
+        try:
+            AddFirewallRules.check_firewalld_rule_applied(self._dst_ip, self._uid)
+        except Exception as error:
+            logger.info(
+                "Check if Firewall rules already applied using firewalld.service failed: {0}".format(ustr(error)))
+            return False
+
+        return True
+
+    def _setup_permanent_firewalld_rules(self):
+        if self.__verify_firewall_rules_enabled():
+            logger.info("Firewall rules already set. No change needed.")
+            return
+
+        logger.info("Firewall rules not added yet, adding them now using firewalld.service")
+        # Add rules if not already set
+        AddFirewallRules.add_firewalld_rules(self._dst_ip, self._uid)
+        logger.info("Successfully added the firewall commands using firewalld.service")
+
+    def __verify_network_setup_service_enabled(self):
+        # Check if the custom service has already been enabled
+        cmd = ["systemctl", "is-enabled", self._network_setup_service_name]
+        try:
+            return shellutil.run_command(cmd).rstrip() == "enabled"
+        except CommandError as error:
+            msg = "{0} not enabled. Command: {1}, ExitCode: {2}\nStdout: {3}\nStderr: {4}".format(
+                self._network_setup_service_name, ' '.join(cmd), error.returncode, error.stdout, error.stderr)
+        except Exception as error:
+            msg = "Ran into error, {0} not enabled. Error: {1}".format(self._network_setup_service_name, ustr(error))
+
+        logger.info(msg)
+        return False
+
+    def _setup_network_setup_service(self):
+        if self.__verify_network_setup_service_enabled():
+            logger.info("Service: {0} already enabled. No change needed.".format(self._network_setup_service_name))
+            self.__log_network_setup_service_logs()
+
+        else:
+            # Create unit file with default values
+            self.__set_service_unit_file()
+            logger.info("Successfully added and enabled the {0}".format(self._network_setup_service_name))
+
+        # Even if service is enabled, we need to overwrite the drop-in file with the current IP in case it changed.
+        # This is to handle the case where WireIP can change midway on service restarts.
```

Comment:
Good point. These parameters also need to be overwritten to update the egg's path during agent auto-update. We should probably add that to the comment as well.

Reply:
Ahh good point, let me add that too.